src/NuGet.Core/NuGet.Protocol/Resources/VulnerabilityInfoResourceV3.cs

```diff
@@ -0,0 +1,168 @@
+        {
+            _sourceRepository = sourceRepository ?? throw new ArgumentNullException(nameof(sourceRepository));
+        }
+
+        internal async Task<IReadOnlyList<V3VulnerabilityIndexEntry>> GetVulnerabilityFilesAsync(SourceCacheContext cacheContext, ILogger log, CancellationToken cancellationToken)
+        {
+            Uri vulnerabilityIndexUrl = await GetIndexUrlAsync(cancellationToken);
+            HttpSourceResource httpSourceResource = await _sourceRepository.GetResourceAsync<HttpSourceResource>(cancellationToken);
+
+            HttpSourceCacheContext httpSourceCacheContext = HttpSourceCacheContext.Create(cacheContext, isFirstAttempt: true);
+            var request = new HttpSourceCachedRequest(vulnerabilityIndexUrl.OriginalString, "vuln_index", httpSourceCacheContext);
+            IReadOnlyList<V3VulnerabilityIndexEntry>? vulnFiles = await httpSourceResource.HttpSource.GetAsync(request,
+                result =>
+            {
+                using (var textReader = new StreamReader(result.Stream))
+                using (var jsonReader = new JsonTextReader(textReader))
+                {
+                    var parsed = JsonExtensions.JsonObjectSerializer.Deserialize<IReadOnlyList<V3VulnerabilityIndexEntry>>(jsonReader);
+                    return Task.FromResult(parsed);
+                }
+            },
+                log,
+                cancellationToken);
+
+            if (vulnFiles == null)
+            {
+                string message = string.Format("Package source '{0}' vulnerability data index '{1}' did not successfully deserialize. This indicates a problem with the JSON file contents returned from the server.",
+                    _sourceRepository.PackageSource.Name,
+                    vulnerabilityIndexUrl.OriginalString);
+                throw new FatalProtocolException(message);
+            }
+
+            return vulnFiles;
+
+            async Task<Uri> GetIndexUrlAsync(CancellationToken cancellationToken)
+            {
+                ServiceIndexResourceV3 serviceIndex = await _sourceRepository.GetResourceAsync<ServiceIndexResourceV3>(cancellationToken);
+                return serviceIndex.GetServiceEntryUri(ServiceTypes.VulnerabilityInfo);
+            }
+        }
+
+        internal async Task<IReadOnlyDictionary<string, IReadOnlyList<VulnerabilityInfo>>> GetVulnerabilityDataAsync(
+            Uri uri,
+            string name,
+            SourceCacheContext cacheContext,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            HttpSourceResource httpSourceResource = await _sourceRepository.GetResourceAsync<HttpSourceResource>(cancellationToken);
+
+            HttpSourceCacheContext httpSourceCacheContext = HttpSourceCacheContext.Create(cacheContext, isFirstAttempt: true);
+            var request = new HttpSourceCachedRequest(uri.OriginalString, "vuln_data_" + name, httpSourceCacheContext);
```

Comment:
I'm a bit concerned using `@name` for the cache key. Per the spec:

> @name, a user friendly name for the page, must be unique.

Does the caching normalize the key? Should we add limits to the `@name` to ensure it cannot break NuGet's cache key logic?

Reply:
I added some limits to prevent malicious (or just careless) servers from causing clients too many problems. Here's a PR to update the spec: https://github.com/NuGet/Home/pull/12529#discussion_r1159438144